command/monitor/monitor_test.go

```diff
@@ -120,3 +122,66 @@ func TestMonitorCommand_LogJSONValidFlag(t *testing.T) {
 		t.Fatal("timed out waiting for exit")
 	}
 }
+
+func TestMonitorCommand_LogJSONValidFormat(t *testing.T) {
+	if testing.Short() {
+		t.Skip("too slow for testing.Short")
+	}
+
+	t.Parallel()
+	a := agent.StartTestAgent(t, agent.TestAgent{})
+	defer a.Shutdown()
+
+	shutdownCh := make(chan struct{})
+
+	ui := cli.NewMockUi()
+	c := New(ui, shutdownCh)
+	args := []string{"-http-addr=" + a.HTTPAddr(), "-log-json"}
+
+	// Buffer it so we don't deadlock when blocking send on shutdownCh triggers
+	// Run to return before we can select on it.
+	exitCode := make(chan int, 1)
+
+	// Run the monitor in another go routine. If this doesn't exit on our "signal"
+	// then the whole test will hang and we'll panic (to not blow up if people run
+	// the suite without -timeout)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		wg.Done() // Signal that this goroutine is at least running now
+		exitCode <- c.Run(args)
+	}()
+
+	// Wait for that routine to at least be running
+	wg.Wait()
+
+	// Read the logs and try to json marshall it
+	go func() {
+		time.Sleep(1 * time.Second)
+		outputs := ui.OutputWriter.String()
+		for count, output := range strings.Split(outputs, "\n") {
+			if output != "" && count > 0 {
+				jsonLog := new(map[string]interface{})
+				err := json.Unmarshal([]byte(output), jsonLog)
+				if err != nil {
+					exitCode <- -1
+				}
+				if len(*jsonLog) <= 0 {
+					exitCode <- 1
+				}
+			}
+		}
+		shutdownCh <- struct{}{}
+
+	}()
+
+	select {
+	case ret := <-exitCode:
+		if ret != 0 {
```

Comment:
There are two goroutines sending to this chan, but this only checks the return value once, which seems like it could easily pass even when one of the goroutines fails.

Instead of running the assertion in a goroutine, could we use a `bufio.Scanner` in the main thread? We could then read lines, and assert some number of log lines is json encoded. Once it successfully we can shutdown the agent and goroutines.

Maybe we'll need to run another goroutine to perform a couple operations to ensure that the agent is producing logs.